### DISCUSSION TITLE
Errors with FLAC 1.4.0 and it's new 32-bit PCM support

### DISCUSSION THREAD

**Discussion by radorn** [enzo1982/freac#374] [Audio Format Discussion]
I've tried to convert some 32-bit integrer PCM files to FLAC with the latest "continuous" build of fre:ac (2022-09-11 exe file date), to try the new 32bit support on the codec.
But the resulting files are unreasonably small and they decode to just rythmic clicks with no resemblance to the original audio, or anything even recognizable.
Of course, built-in verification routines confirm the output file doesn't compare to the input.

**Comment by enzo1982** [comment]
Thanks for reporting this!

There's a new continuous build now that should fix this. Routines for converting between fre:ac and FLAC buffer formats were missing for the 32 bit case.

**Reply by radorn** [reply]
Thank you for maintaining the software!

**Comment by radorn** [comment]
I've got around to try the new build. 32bit flac works now. Thanks.
I just found a little non-critical weirdness now.
I started using fre:ac for a way to convert a bunch of 32bit integer wavs to wavpack, since foobar, with it's internal 32bit floating point signal path, wouldn't preserve the full 32bit integer PCM. So I used fre:ac to convert these WAVs to WavPack.
Now that you've fixed this issue, I tried converting those to FLAC, and it worked, but, for some files, fre:ac warns me that it thinks I'm converting from lossy to lossless. It's not all of them, just a few so far, which I've had to narrow down manually. The thing is those are all entirely lossless WavPack files, encoded at the same time and with the same settings as all the others that don't trigger the warning. They're not even encoded in hybrid mode, just regular lossless.

There's no warning if I try to convert these WavPack files to WAV. There's only a warning when I try converting them to other lossless formats (WavPack, FLAC, Monkey's, TAK...). And, again, it's not for all files, but just those few. I can't tell what may be different about them to make fre:ac think they are lossy.

I can try to upload them somewhere, if you need them, as they aren't big at all. In fact, I considered if maybe it was the small size that triggered this, but there are other smaller files that don't trigger the warning, so I have no idea what's going on.

**Reply by enzo1982** [reply]
When dealing with formats that support lossy and lossless modes and fre:ac does not definitely know which one it is, it makes a guess based on the file's bitrate.

For that, fre:ac takes the raw bitrate of the sample format, i.e. `numberOfChannels * sampleRate * bitsPerSample`, and compares it to the compressed bitrate of the file. If the compressed bitrate is less than 1/3rd of the raw bitrate, the file is considered to probably be lossy.

This can of course go wrong, especially if a file contains a lot of silence or has been upscaled (i.e. converted from 16 or 24 bit to 32 bit).

This affects only formats that use external decoders and can be either lossy or lossless. So basically only WavPack and OptimFROG are affected.

**Reply by radorn** [reply]
Oh, I see. You need to resort to that probabilistic approach because the external decoder protocol isn't designed to communicate the particular encoding method that hybrid codecs can have. Thanks for explaining.
I'll just ignore or disable the warning, then.

**Comment by radorn** [comment]
Sorry to keep bothering you with stuff. But I was converting the WavPacks I previously encoded from WAV with the stable version of fre:ac, now to FLAC 32-bit with the new continuous build. When it finished, I did a search to compare total size (as I noticed that FLAC now compresses a little bit more overall), and this is when I noticed there were four files missing. So I hunted them down and tried to load them, but nor fre:ac itself or foobar2000 recognize them.
I no longer have the original WAVs from which these were produced, so I'm not sure what to do.
I always use verification and fre:ac didn't warn me about them when I encoded (not that I remember, anyway). Not sure what the original WAVs had, but fre:ac did accept the WAVs and encoded them to WV and didn't find anything wrong with them once done, but now it refuses to even load these files it previously produced and gave the OK to.

In a hex editor they look to me like bare headers with metadata and no audio data.
